### PR TITLE
FINERACT-2107: Interest Refund - Partial Refund cases

### PR DESCRIPTION
**Pull request by somasorosdpc**
## Description

[FINERACT-2107](https://issues.apache.org/jira/browse/FINERACT-2107): Interest Refund - Partial Refund cases

Interest Refund calculation on Payout Refund and Merchant Issued Refund if configured for interest bearing progressive loans. + Integration tests
** covering partial refund in single transaction.

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [x] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [x] Create/update unit or integration tests for verifying the changes made.

- [x] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [x] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [x] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/ProgressiveLoanInterestRefundServiceImpl.java [hunk 1]
```diff
@@ -54,53 +54,48 @@ public boolean canHandle(Loan loan) {
...
     }
 
-    private static boolean omitDisbursements(LoanTransaction lt, final AtomicReference<BigDecimal> refundFinal) {
+    private static void simulateRepaymentForDisbursements(LoanTransaction lt, final AtomicReference<BigDecimal> refundFinal,
+            List<LoanTransaction> collect) {
+        collect.add(lt);
         if (lt.getTypeOf().isDisbursement() && refundFinal.get().compareTo(BigDecimal.ZERO) > 0) {
             if (lt.getAmount().compareTo(refundFinal.get()) <= 0) {
+                collect.add(
+                        new LoanTransaction(lt.getLoan(), lt.getLoan().getOffice(), REPAYMENT.getValue(), lt.getDateOf(), lt.getAmount(),
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/ProgressiveLoanInterestRefundServiceImpl.java:63]: Can you please tell me about this? I thought we were just amending the disbursement and calculate interest based on the remaining amount.

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanInterestRefundTest.java [hunk 2]
```diff
@@ -432,6 +435,285 @@ public void verifyUC05() {
...
+            PostLoansLoanIdTransactionsResponse postLoansLoanIdTransactionsResponse = loanTransactionHelper
+                    .makeLoanRepayment("PayoutRefund", "22 January 2021", 500F, loanId.intValue());
+            Assertions.assertNotNull(postLoansLoanIdTransactionsResponse);
+            Assertions.assertNotNull(postLoansLoanIdTransactionsResponse.getResourceId());
+
+            logTransactions(loanId);
+            verifyTransactions(loanId, transaction(250.0, "Disbursement", "01 January 2021"), //
+                    transaction(750.0, "Disbursement", "01 January 2021"), //
+                    transaction(500.0, "Payout Refund", "22 January 2021"), //
+                    // transaction(5.75, "Accrual", "22 January 2021"), //
```
Comment (galovics) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanInterestRefundTest.java]: Comment.

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanInterestRefundTest.java [hunk 3]
```diff
@@ -432,6 +435,285 @@ public void verifyUC05() {
...
+            PostLoansLoanIdTransactionsResponse postLoansLoanIdTransactionsResponse = loanTransactionHelper
+                    .makeLoanRepayment("PayoutRefund", "22 January 2021", 500F, loanId.intValue());
+            Assertions.assertNotNull(postLoansLoanIdTransactionsResponse);
+            Assertions.assertNotNull(postLoansLoanIdTransactionsResponse.getResourceId());
+
+            logTransactions(loanId);
+            verifyTransactions(loanId, transaction(250.0, "Disbursement", "01 January 2021"), //
+                    transaction(750.0, "Disbursement", "07 January 2021"), //
+                    transaction(500.0, "Payout Refund", "22 January 2021"), //
+                    // transaction(2.46, "Accrual", "22 January 2021"), //
```
Comment (galovics) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanInterestRefundTest.java]: Comment.

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanInterestRefundTest.java [hunk 4]
```diff
@@ -432,6 +435,285 @@ public void verifyUC05() {
...
+                    transaction(171.29, "Repayment", "01 April 2021"), //
+                    transaction(171.29, "Repayment", "01 May 2021"), //
+                    transaction(171.29, "Repayment", "01 June 2021"), //
+                    transaction(171.32, "Repayment", "01 July 2021"), //
+                    transaction(500.0, "Payout Refund", "11 July 2021"), //
+                    transaction(20.41, "Interest Refund", "11 July 2021")); //
+        });
+    }
+
+    // @Test
```
Comment (galovics) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanInterestRefundTest.java]: Comment.

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanInterestRefundTest.java [hunk 5]
```diff
@@ -432,6 +435,285 @@ public void verifyUC05() {
...
+            logTransactions(loanId);
+            verifyTransactions(loanId, transaction(1000.0, "Disbursement", "01 January 2021"), //
+                    transaction(500.0, "Merchant Issued Refund", "11 January 2021"), //
+                    transaction(1.37, "Interest Refund", "11 January 2021"), //
+                    transaction(500.0, "Payout Refund", "22 January 2021"), //
+                    transaction(2.87, "Interest Refund", "22 January 2021")); //
+        });
+    }
+
+    private void logLoanDetails(GetLoansLoanIdResponse loanDetails) {
```
Comment (galovics) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanInterestRefundTest.java]: This doesn't only log, please name the method appropriately.

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanInterestRefundTest.java [hunk 6]
```diff
@@ -432,6 +435,285 @@ public void verifyUC05() {
...
+
+    @Test
+    public void verifyUC07() {
+        AtomicReference<Long> loanIdRef = new AtomicReference<>();
+        runAt("1 January 2021", () -> {
+            PostLoanProductsResponse loanProduct = loanProductHelper
+                    .createLoanProduct(create4IProgressive().daysInMonthType(DaysInMonthType.ACTUAL) //
+                            .daysInYearType(DaysInYearType.ACTUAL) //
+                            .supportedInterestRefundTypes(new ArrayList<>()).addSupportedInterestRefundTypesItem("PAYOUT_REFUND") //
+                            .addSupportedInterestRefundTypesItem("MERCHANT_ISSUED_REFUND") //
```
Comment (galovics) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanInterestRefundTest.java]: Would it be possible to create constants for these static values? Would be much better for maintainability.

**Review by galovics — CHANGES_REQUESTED** [review]

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanInterestRefundTest.java (hunk 4 above)
Comment (somasorosdpc) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanInterestRefundTest.java]: I removed the unused test case too

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanInterestRefundTest.java (hunk 6 above)
Comment (somasorosdpc) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanInterestRefundTest.java]: I've created a static class in Base class for options, and removed the unnecessary `.supportedInterestRefundTypes(new ArrayList<>())`

**Pull request merged**